Refactor analyzepsbt for use outside RPC code

gwillen opened the pull request:
Refactor the analyzepsbt RPC into (1) an AnalyzePSBT function, which returns
its output as a new strongly-typed PSBTAnalysis struct, and (2) a thin wrapper
which converts the struct into a UniValue for RPC use.

----

As with my previous refactoring PR, I need this because I am creating a dependency on this code from the GUI. Per discussion in #bitcoin-core-dev on IRC, since we don't want to create a dependency on UniValue in anything outside RPC, I introduced some new structs to hold the info we get when analyzing a PSBT. For the field types, I used whatever types are already used internally for this data (e.g. CAmount, CFeeRate, CKeyID), and only convert to int/string etc. in the wrapper.

@achow101, maybe take the first look? :-)

gwillen committed:
Refactor analyzepsbt for use outside RPC code

Refactor the analyzepsbt RPC into (1) an AnalyzePSBT function, which returns
its output as a new strongly-typed PSBTAnalysis struct, and (2) a thin wrapper
which converts the struct into a UniValue for RPC use.

fanquake commented:
[Travis](https://travis-ci.org/bitcoin/bitcoin/jobs/500229801):
```
A new circular dependency in the form of "core_io -> psbt -> core_io" appears to have been introduced.
^---- failure generated from test/lint/lint-circular-dependencies.sh
```

gwillen commented:
Hmm, that's not great, is it. I guess I forgot to run lint-all, sorry. (I have to raise the file descriptor limit before running, or it crashes on my mac, so it's easy to forget...)

I am a little confused about this lint, because it appears I didn't introduce a circular header dependency -- rather, core_io.h (the interface) depends on psbt.h, but psbt._cpp_ (the implementation) depends on core_io.h. This doesn't seem especially offensive to me, since it's not actually circular? Is it intentional that the lint flags this situation?

gwillen commented:
... but apparently the dependency was no longer needed anyway, so I don't need to litigate this. ;-) Fixed!

practicalswift commented on src/psbt.cpp:
```diff
@@ -325,3 +330,139 @@ TransactionError CombinePSBTs(PartiallySignedTransaction& out, const std::vector
+        bool success = true;
+
+        for (unsigned int i = 0; i < psbtx.tx->vin.size(); ++i) {
+            PSBTInput& input = psbtx.inputs[i];
```
The scope of `input` could be reduced here without any change in behaviour, right?

gwillen replied:
Yes, that's true -- would you prefer I did? I normally agree about "less scope is better", but here that instinct is clashing with my instinct to keep parallel things together, and definitions at the top of the relevant block.

practicalswift replied:
Thanks for the explanation. Makes sense!

promag commented on src/psbt.h:
```diff
@@ -580,4 +610,12 @@ bool FinalizeAndExtractPSBT(PartiallySignedTransaction& psbtx, CMutableTransacti
+ * @param[in] psbtx the PSBT to analyze
+ * @return A PSBTAnalysis with information about the provided PSBT.
+ */
+PSBTAnalysis AnalyzePSBT(PartiallySignedTransaction psbtx);
```
`const PartiallySignedTransaction& psbtx`?

gwillen replied:
I tried, but I had to back it out, because of the pattern of using "SignPSBTInput(DUMMY_SIGNING_PROVIDER, ...)" to get signature data without modifying the PSBT. So we either have to take the psbt by value, or by non-const reference.

We could get around this with a const cast, but I'm against it unless we think the cost of copying a PartiallySignedTransaction is unacceptable.

promag replied:
Makes sense to keep by value. The code could be refactored but at the end each input would have to be copied to be able to call `FillSignatureData()` and `FromSignatureData`.

promag commented on src/psbt.cpp:
```diff
@@ -325,3 +330,139 @@ TransactionError CombinePSBTs(PartiallySignedTransaction& out, const std::vector
+        for (unsigned int i = 0; i < psbtx.tx->vin.size(); ++i) {
+            PSBTInput& input = psbtx.inputs[i];
+            if (SignPSBTInput(DUMMY_SIGNING_PROVIDER, psbtx, i, 1, nullptr, true)) {
+                mtx.vin[i].scriptSig = input.final_script_sig;
```
Could move this after `GetInputUTXO()`?

gwillen replied:
I'd be happy to, but I'm not sure I understand why? Can you elaborate?

promag replied:
consider nit, just one loop exit:
```cpp
if (!SignPSBTInput(...) || !psbtx.GetInputUTXO(...)) {
     success = false
     break;
}
```

gwillen replied:
Done.

promag replied:
nit, could drop `else {` block or invert conditions:
```cpp
for (unsigned int i = 0; i < psbtx.tx->vin.size(); ++i) {
    PSBTInput& input = psbtx.inputs[i];
    Coin newcoin;

    if (SignPSBTInput(DUMMY_SIGNING_PROVIDER, psbtx, i, 1, nullptr, true) && psbtx.GetInputUTXO(newcoin.out, i)) {
        mtx.vin[i].scriptSig = input.final_script_sig;
        mtx.vin[i].scriptWitness = input.final_script_witness;
        newcoin.nHeight = 1;
        view.AddCoin(psbtx.tx->vin[i].prevout, std::move(newcoin), true);
    } else {
        success = false;
        break;
    }
}
```

gwillen replied:
At this point I think I better leave it as Future Work.

promag commented on src/psbt.cpp:
```diff
@@ -325,3 +330,139 @@ TransactionError CombinePSBTs(PartiallySignedTransaction& out, const std::vector
+
+std::string PSBTRoleName(PSBTRole role) {
+    switch (role) {
+        case PSBTRole::UPDATER:   return "updater";
```
I think we align `case` with `switch`. nit, I'd ignore `return`s alignment.

gwillen replied:
Done.

gwillen commented:
Hmm, I seem to have introduced a linker problem that does not happen on OS X. I will get around to looking at it in a Linux VM, but if someone could look at it and go "oh, I know what you did wrong", I'm sure it would save me a bunch of time. ;-)

sipa commented:
psbt.cpp is in libbitcoin_common, and consensus/tx_verify.cpp is in libbitcoin_server. libbitcoin_common in linked into `bitcoin-tx`, but `libbitcoin_server` isn't.

gwillen committed:
Move PSBT decoding functions from core_io to psbt.cpp

Move PSBT decoding functions from core_io.h/core_read.cpp to psbt.h/psbt.cpp,
to deal with a linker issue.

gwillen commented:
Thanks @sipa.

The easiest solution seems -- after whipping out a linux machine and doing a lot of repeatedly fiddling with Makefile.am and rebuilding everything from scratch -- to be moving psbt.cpp from libbitcoin_common to libbitcoin_server. LMK what you think of that plan.

Unfortunately the tree of dependencies from AnalyzePSBT into libbitcoin_server runs very deep. The chain of dependencies is mostly fake -- of the form "X coincidentally exists in the same file as something that depends on Y" -- but still seems annoying to break.

The actual issue is not with _most_ of the functionality of AnalyzePSBT, but just with the ancillary bit at the end which computes the size of a transaction, in order to compute the feerate. So one other option would be to break that bit out into a separate function and throw it into libbitcoin_server somewhere.

ryanofsky commented on src/psbt.h:
```diff
@@ -548,8 +550,36 @@ struct PartiallySignedTransaction
+    EXTRACTOR
+};
+
+struct PSBTInputAnalysis {
```
It'd be nice to have some documentation for these structs. There could be a comment saying struct fields correspond to fields returned by the `analyzepsbt` method, if you didn't want to just copy the documentation from `analyzepsbt`

gwillen replied:
Nah, thanks, this is a good point, I may as well document them properly.

ryanofsky reviewed (approved):
utACK 5f96bf29225d7943314d68e10e19ee2fe3ccf3f3. I left a suggestion, but think this looks good as is. I just compared the moved code to make sure behavior did not appear to be changing.

gwillen committed:
Add documentation of struct PSBTAnalysis et al

gwillen commented:
Ok, added documentation, and did the refactor requested by @promag. Sorry @ryanofsky, if you wouldn't mind glancing at the minor code change and re-acking. I am convinced it does not change behavior.

gwillen commented:
Travis failure is weird:

```Downloading archive: https://language-archives.travis-ci.com/python/binaries/ubuntu/16.04/x86_64/python-3.5.tar.bz2
0.20s$ curl -sSf -o python-3.5.tar.bz2 ${archive_url}
curl: (22) The requested URL returned error: 403 Forbidden
Unable to download 3.5 archive. The archive may not exist. Please consider a different version.```

sipa commented:
It's really unfortunate that the PSBT code needs to be in libbitcoin_server. I'm thinking of creating a separate PSBT signer tool using the signer/psbt source code, but with it in libbitcoin_server, that tool will need to include the whole networking, RPC, and validation logic.

Is it depending on anything more than `GetVirtualTransactionSize`? If it is, I would suggest just moving `GetTransactionWeight` and friends from src/consensus/validation.h, and `GetVirtualTransactionSize` and friends from src/policy/policy.{h,cpp} to src/primitives/{transaction,block} respectively.

gwillen commented:
@sipa I'm happy to give it a shot.

From my notes, there were two offenders:
- GetVirtualTransactionSize, as noted, which lives in policy.cpp
- GetTransactionSigOpCost, which lives in consensus/tx_verify.cpp

The former seems easy to deal with -- it's just a small amount of math involving a couple of constants. 

The latter might also be relatively standalone -- the file it's in has deeper dependencies spidering into libbitcoin_server, but I couldn't tell if they were real dependencies of GetTransactionSigOpCost or if they were coincidentally being dragged in by something else.

Do you have a proposed place to move GetTransactionSigOpCost?

gwillen commented:
(The other option would be to remove the size and feerate calculations from AnalyzePSBT, or move just those elsewhere. They're pretty standalone.)

sipa commented:
I have a simpler solution: replace the `GetTransactionSigOpCost()` call with `0` (which means the whole call can just become `GetVirtualTransactionSize(ctx)`, and only that function needs to move elsewhere).

To explain why this is safe: reason why GetVirtualTransactionSize takes in a sigop count as argument is because in the mempool/feerate/mining code we use vsize = max(actual_vsize(tx), sigops(tx)*someconstant). This brings everything to a single unit we can optimize for, but still guarantees that transactions with excessively high amount of signature validation operations won't cause the block-wide sigops limit to be exceeded.

However, the wallet/signing/psbt side of things doesn't need to care about this. They will never deal with transactions where the sigop side of that max dominates (it would require more signature checks than 1 per 20 bytes of witness data).

ryanofsky reviewed (approved):
utACK 9d98086f22733cd8363d3b1df022d6e7f7dfd626. Changes since last review: changing if/else order not assigning empty fields, adding documentation as suggested in previous comments, also changing int to CAmount.

gwillen commented:
ok, my apologies to @ryanofsky and I appreciate your patience, this version has exactly two changes:

- Rebase over #15559, which changes the format of the estimated feerate from a string to an integer in the RPC output (does not affect the output of the AnalyzePSBT function itself, only how it's formatted in the wrapper in rpc/rawtransaction.cpp)
- Per @sipa 's request, revert the move into libbitcoin_server, putting psbt.cpp back in libbitcoin_common. Fingers crossed that Travis likes this, but it appears experimentally that moving the PSBT-related functions from core_io into psbt is enough to satisfy the linker, without actually moving PSBT around.

(This does not quite actually solve his problem; the reason the linker accepts this is that bitcoin-tx has no dependency on psbt.cpp, and the corresponding object file is now being dropped from the link, so the fact that it has in turn a broken dependency never comes up. Actually _using_ psbt without linking bitcoin_server will still give a link error. However, it will be _much_ easier to fix this once the code that is using psbt actually exists, because right now it's very hard to get the linker not to optimize out the problem, so it's really hard to confidently fix it.)

ryanofsky reviewed (approved):
utACK deef14c4f5d6c997e5cc68fc4007104dc2051606. Only changes since last review are leaving psbt.cpp in libbitcoin_common (instead of libbitcoin_server) as described above and rebasing after conflict with #15559

gwillen commented:
Definitely hoping to get this merged soon. @sip or @promag, maybe take a look and see if you have anything further? :-)

sipa commented on src/core_read.cpp:
```diff
@@ -177,33 +177,6 @@ bool DecodeHexBlk(CBlock& block, const std::string& strHexBlk)
     return true;
 }
 
-bool DecodeBase64PSBT(PartiallySignedTransaction& psbt, const std::string& base64_tx, std::string& error)
```
I think this means you can remove the `#include <psbt.h>` from this file.

gwillen replied:
yep, done

sipa commented on src/core_io.h:
```diff
@@ -37,11 +37,6 @@ bool DecodeHexBlockHeader(CBlockHeader&, const std::string& hex_header);
  */
 bool ParseHashStr(const std::string& strHex, uint256& result);
 std::vector<unsigned char> ParseHexUV(const UniValue& v, const std::string& strName);
-
```
I think this means you can remove the `struct PartiallySignedBitcoinTransaction` forward declare from this file.

gwillen replied:
yep, done

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1895,148 +1895,56 @@ UniValue analyzepsbt(const JSONRPCRequest& request)
-    CAmount in_amt = 0;
-    for (unsigned int i = 0; i < psbtx.tx->vin.size(); ++i) {
-        PSBTInput& input = psbtx.inputs[i];
+    for (auto input : psbta.inputs) {
```
`const auto& input` to avoid copies.

gwillen replied:
done

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1895,148 +1895,56 @@ UniValue analyzepsbt(const JSONRPCRequest& request)
-        only_missing_sigs = false;
-        result.pushKV("next", "extractor");
+    if (psbta.estimated_vsize != nullopt) {
+        result.pushKV("estimated_vsize", (int)*psbta.estimated_vsize);
```
Is this int cast necessary?

gwillen replied:
According to the compiler, some cast is required:

```
rpc/rawtransaction.cpp:1938:16: error: call to member function 'pushKV' is ambiguous
        result.pushKV("estimated_vsize", *psbta.estimated_vsize);
        ~~~~~~~^~~~~~
./univalue/include/univalue.h:124:10: note: candidate function
    bool pushKV(const std::string& key, int64_t val_) {
         ^
./univalue/include/univalue.h:128:10: note: candidate function
    bool pushKV(const std::string& key, uint64_t val_) {
         ^
./univalue/include/univalue.h:132:10: note: candidate function
    bool pushKV(const std::string& key, bool val_) {
         ^
./univalue/include/univalue.h:136:10: note: candidate function
    bool pushKV(const std::string& key, int val_) {
         ^
./univalue/include/univalue.h:140:10: note: candidate function
    bool pushKV(const std::string& key, double val_) {
         ^
./univalue/include/univalue.h:115:10: note: candidate function
    bool pushKV(const std::string& key, const UniValue& val);
```

sipa replied:
Looks like it's needed in that case.

sipa reviewed (commented):
utACK deef14c4f5d6c997e5cc68fc4007104dc2051606, only nits

gwillen commented:
@sipa Thanks, all your comments addressed.

sipa commented:
utACK 892eff05f115c0b002d0e0b6ffc3ab418480d25c

sipa commented:
Feel like reviewing, @achow101?

ryanofsky reviewed (approved):
utACK 892eff05f115c0b002d0e0b6ffc3ab418480d25c. Just small cleanups since the last review: removing unneeded include, forward decl, adding const ref

achow101 commented:
utACK 892eff05f115c0b002d0e0b6ffc3ab418480d25c
